infrastructure we need for spl-token and calls to rust contracts (#794)
hyperledger-labs/solang: docs/language.rst

    import * as defs from "defines.sol";

Builtin Imports
_______________

> Some builtin functionality only available after importing. The following structs

Reviewer: ```suggestion
Some builtin functionality is only available after importing. The following structs
```

Author: Thanks for spotting that. Good reviews, I appreciate it!

that it must be done on a contract type variable. Any error in an external call can
be handled with :ref:`try-catch`.

Internal calls and externals calls
___________________________________

> An internal function call is one that is executed by the current contract. This
> is much more efficient than an external call. An external call requires the
> address of the contract to call, the arguments must be *abi encoded* (also known
> as serialization), then the runtime must set up the VM for the called contract
> (the callee). It must decode the arguments, and then encode return values. Lastly
> the first contract (the caller) must decode return values.

Reviewer: ```suggestion
An internal function call is executed by the current contract. This
is much more efficient than an external call, which requires the
address of the contract to call, whose arguments must be *abi encoded* (also known
as serialization). Then, the runtime must set up the VM for the called contract
(the callee), decode the arguments, and encode return values. Lastly,
the first contract (the caller) must decode return values.
```

Author: All good, I've amened the commit with your suggestion

> Whenever a call is done on a contract type, this will intiate an external call.

Reviewer: ```suggestion
A call done on a contract type will always initiate an external call.
```

Author: All good, I've amened the commit with your suggestion

Note that ``this`` returns the current contract, so ``this.foo()`` will do an
external call, which is much more expensive than ``foo()``.

> Passing accounts on Solana on external calls

Reviewer: ```suggestion
Passing accounts on Solana during external calls
```
I just wanted to avoid the repetition of `on`. Feel free to change the preposition, if `during` does not sound good.

Author: changed it to with

____________________________________________

The Solana runtime allows you the specify the accounts to be passed for an

> external call. This is specified in an array of the struct ``AccountMeta``. This

Reviewer: Is `AccountMeta` going to be a reserved word?

Reviewer: We could do something like `SolanaSdk.AccountMeta`.

Author: Now you've made `SolanaSdk` a reserved word (and it's really ugly).

Author: So I've decided to bite the bullet and go ahead with a change I had planned for some time.

If we want to introduce new types, or builtin functions, we should hide them using solidity imports. The import system used by solidity mimics what javascript does. In javascript/node, builtins (apart from ancient onces) are hidden behind imports.

I've change the code and now there is an import called:

```
import 'solana';
```
Which imports `AccountMeta` and `AccountInfo`. You can do all the renaming tricks that import permits.

This keeps the namespace clean. 

On Solana we also have a `verifySignature()` call that should go into this import (something for another PR)

that it must be done on a contract type variable. Any error in an external call can
be handled with :ref:`try-catch`.

Internal calls and externals calls
___________________________________

An internal function call is executed by the current contract. This
is much more efficient than an external call, which requires the
address of the contract to call, whose arguments must be *abi encoded* (also known
as serialization). Then, the runtime must set up the VM for the called contract
(the callee), decode the arguments, and encode return values. Lastly,
the first contract (the caller) must decode return values.

A method call done on a contract type will always be an external call.
Note that ``this`` returns the current contract, so ``this.foo()`` will do an
external call, which is much more expensive than ``foo()``.

Passing accounts with external calls on Solana
______________________________________________

The Solana runtime allows you the specify the accounts to be passed for an
external call. This is specified in an array of the struct ``AccountMeta``, which
has to be imported from the builtin file ``'solana'``. This import is builtin when

> you build your Solidity for Solana.

Reviewer: It is a nice idea do add a section in the docs to talk about the contents of this 'solana' file.

Reviewer: Specify the structs available, their attributes and the latter's types.

Author: I've introduced a builtin imports section.
